integration/ruler_test.go

```diff
@@ -1038,6 +1038,168 @@ func TestRulerDisablesRuleGroups(t *testing.T) {
+	ts := time.Now()
+	require.NoError(t, c.SetRuleGroup(alertRuleWithKeepFiringFor(groupName, ruleName, expression, model.Duration(10*time.Second)), namespace))
+	// Wait until rule group has tried to evaluate the rule.
+	require.NoError(t, ruler.WaitSumMetricsWithOptions(e2e.GreaterOrEqual(2), []string{"cortex_prometheus_rule_evaluations_total"}, e2e.WithLabelMatchers(m), e2e.WaitMissingMetrics))
+
+	updatedGroups, err := c.GetPrometheusRules(e2ecortex.RuleFilter{
+		RuleNames: []string{ruleName},
+	})
+	require.NoError(t, err)
+	require.NotEmpty(t, updatedGroups)
+	require.Equal(t, 1, len(updatedGroups[0].Rules))
+
+	alert = parseAlertFromRule(t, updatedGroups[0].Rules[0])
+	require.Equal(t, "firing", alert.State)
+	require.Equal(t, float64(10), alert.KeepFiringFor)
+	require.Equal(t, 1, len(alert.Alerts))
+	require.NotEmpty(t, alert.Alerts[0].KeepFiringSince)
+	require.Greater(t, alert.Alerts[0].KeepFiringSince.UnixNano(), ts.UnixNano(), "KeepFiringSince value should be after expression is resolved")
+
+	time.Sleep(10 * time.Second) // Sleep beyond keepFiringFor time
+	updatedGroups, err = c.GetPrometheusRules(e2ecortex.RuleFilter{
+		RuleNames: []string{ruleName},
+	})
+	require.NoError(t, err)
+	require.NotEmpty(t, updatedGroups)
+	require.Equal(t, 1, len(updatedGroups[0].Rules))
+	alert = parseAlertFromRule(t, updatedGroups[0].Rules[0])
+	require.Equal(t, 0, len(alert.Alerts)) // alert should be resolved once keepFiringFor time expires
+}
+
+func parseAlertFromRule(t *testing.T, rules interface{}) *alertingRule {
+	responseJson, err := json.Marshal(rules)
+	require.NoError(t, err)
+
+	alertResp := &alertingRule{}
+	require.NoError(t, json.Unmarshal(responseJson, alertResp))
+	return alertResp
+}
+
+type alertingRule struct {
+	// State can be "pending", "firing", "inactive".
+	State          string        `json:"state"`
+	Name           string        `json:"name"`
+	Query          string        `json:"query"`
+	Duration       float64       `json:"duration"`
+	KeepFiringFor  float64       `json:"keepFiringFor"`
+	Labels         labels.Labels `json:"labels"`
+	Annotations    labels.Labels `json:"annotations"`
+	Alerts         []*Alert      `json:"alerts"`
+	Health         string        `json:"health"`
+	LastError      string        `json:"lastError"`
+	Type           v1.RuleType   `json:"type"`
+	LastEvaluation time.Time     `json:"lastEvaluation"`
+	EvaluationTime float64       `json:"evaluationTime"`
+}
+
+// Alert has info for an alert.
+type Alert struct {
+	Labels          labels.Labels `json:"labels"`
+	Annotations     labels.Labels `json:"annotations"`
+	State           string        `json:"state"`
+	ActiveAt        *time.Time    `json:"activeAt"`
+	KeepFiringSince *time.Time    `json:"keepFiringSince,omitempty"`
+	Value           string        `json:"value"`
+}
+
+func alertRuleWithKeepFiringFor(groupName string, ruleName string, expression string, keepFiring model.Duration) rulefmt.RuleGroup {
```

Comment:
I'd love to see we don't duplicate code for this since it is almost the same functionality for `ruleGroupWithRule`. 
But it is not a blocker. We can address later to use builder pattern.